boards/BMS/Test/Src/imd_test.cpp

```diff
@@ -0,0 +1,267 @@
+#include "fff.h"
+#include <gtest/gtest.h>
+#include <utility>
+
+extern "C"
+{
+    DEFINE_FFF_GLOBALS;
+#include "imd_fff.h"
+#include "App_Imd.h"
+}
+
+class Imd_Test : public testing::Test
+{
+  protected:
+    void SetUp() override
+    {
+        imd = App_Imd_Create(
+            get_pwm_frequency, 2.0f, get_pwm_duty_cycle,
+            get_seconds_since_power_on);
+
+        RESET_FAKE(get_pwm_frequency);
+        RESET_FAKE(get_pwm_duty_cycle);
+        RESET_FAKE(get_seconds_since_power_on);
+    }
+
+    void TearDown() override
+    {
+        ASSERT_TRUE(imd != NULL);
+        App_Imd_Destroy(imd);
+        imd = NULL;
+    }
+
+    void SetPwmFrequencyTolerance(float tolerance)
+    {
+        ASSERT_TRUE(imd != NULL);
+        App_Imd_Destroy(imd);
+        imd = App_Imd_Create(
+            get_pwm_frequency, tolerance, get_pwm_duty_cycle,
+            get_seconds_since_power_on);
+    }
+
+    void SetImdCondition(enum Imd_Condition condition)
```

Comment:
@garethellis0  Had to write this function twice, once in `imd_test.cpp` and once in `state_machine_tests.cpp`. Is there an elegant way to re-use it?

Reply:
Yep, you can define a class `BmsTest` that extends `public testing::Test`, then both `ImdTest` and `StateMachineTest` can extend `public BmsTest`. Then you can put shared code in `BmsTest`.

Reply:
Played around with this slightly but unsure  about a couple things

Points I considered:

- Should `struct Imd* imd` just live in `BmsTest`? Or should `StateMachineTest` and `ImdTest` hold its own copy of `sturct Imd* imd`
- Depending on the last point, I would also need to change the function signature for `SetImdCondition`

```
// If "struct Imd* imd" lives in class BmsTest
void BmsTest::SetImdCondition(enum Imd_Condition condition);

// If StateMachineTest and ImdTest hold their own "struct Imd* imd"
void BmsTest::SetImdCondition(struct Imd* imd, enum Imd_Condition condition)
```
- It gets a little more annoying than that actually because `SetImdCondition` needs to modify `{fff_functio}.return_val`, so the most correct function signatures would be 
``` 
// If "struct Imd* imd" lives in class BmsTest
void BmsTest::SetImdCondition(float& fake_frequency_return_value, enum Imd_Condition condition);

// If StateMachineTest and ImdTest hold their own "struct Imd* imd"
void BmsTest::SetImdCondition(struct Imd* imd, float& fake_frequency_return_value, enum Imd_Condition condition)
```

Anyway, I wrote up a [quick draft](https://github.com/UBCFormulaElectric/Consolidated-Firmware/pull/555/commits/ee105058bba125775e424cee9ab2feacbdb35330) that works. Feel free to leave any comments there.

Reply:
Looks good to me!